File: backend/app/modules/income/distribution_service.py
"""
Income Distribution Service.
Handles distribution rules for splitting income across accounts and goals.
"""
from datetime import datetime
from decimal import Decimal
from typing import Optional, List, Tuple
from uuid import UUID
import logging

from sqlalchemy import select, and_
from sqlalchemy.ext.asyncio import AsyncSession

from app.modules.income.models import (
    IncomeDistributionRule,
    IncomeSource,
    IncomeTransaction,
    IncomeTransactionStatus,
    DistributionType,
)
from app.modules.income.schemas import (
    IncomeDistributionRuleCreate,
    IncomeDistributionRuleUpdate,
    IncomeDistributionRuleResponse,
    DistributionPreview,
    IncomeDistributionPreviewResponse,
)
from app.modules.savings.models import SavingsAccount
from app.modules.savings.transaction_service import TransactionService
from app.modules.goals.models import Goal

logger = logging.getLogger(__name__)


class DistributionServiceError(Exception):
    """Base exception for distribution service errors."""
    pass


class RuleNotFoundError(DistributionServiceError):
    """Raised when a distribution rule is not found."""
    pass


class InvalidRuleError(DistributionServiceError):
    """Raised when rule configuration is invalid."""
    pass


class DistributionService:
    """Service for managing income distribution rules."""

    def __init__(self, db: AsyncSession):
        self.db = db

    async def create_rule(
        self,
        user_id: UUID,
        rule_data: IncomeDistributionRuleCreate,
    ) -> IncomeDistributionRule:
        """
        Create a new distribution rule.

        Args:
            user_id: User UUID
            rule_data: Rule creation data

        Returns:
            Created IncomeDistributionRule

        Raises:
            InvalidRuleError: If rule configuration is invalid
        """
        # Validate rule has at least one target
        if not rule_data.target_account_id and not rule_data.target_goal_id:
            raise InvalidRuleError("Rule must have either a target account or target goal")

        # Validate distribution type has required field
        if rule_data.distribution_type == DistributionType.PERCENTAGE:
            if rule_data.percentage is None:
                raise InvalidRuleError("Percentage type requires percentage value")
        elif rule_data.distribution_type == DistributionType.FIXED_AMOUNT:
            if rule_data.amount is None:
                raise InvalidRuleError("Fixed amount type requires amount value")

        # Validate income source exists if specified
        if rule_data.income_source_id:
            source = await self.db.get(IncomeSource, rule_data.income_source_id)
            if not source or source.user_id != user_id:
                raise InvalidRuleError("Invalid income source")

        # Validate target account exists if specified
        if rule_data.target_account_id:
            account = await self.db.get(SavingsAccount, rule_data.target_account_id)
            if not account or account.user_id != user_id:
                raise InvalidRuleError("Invalid target account")

        # Validate target goal exists if specified
        if rule_data.target_goal_id:
            goal = await self.db.get(Goal, rule_data.target_goal_id)
            if not goal or goal.user_id != user_id:
                raise InvalidRuleError("Invalid target goal")

        rule = IncomeDistributionRule(
            user_id=user_id,
            income_source_id=rule_data.income_source_id,
            target_account_id=rule_data.target_account_id,
            target_goal_id=rule_data.target_goal_id,
            distribution_type=rule_data.distribution_type,
            amount=rule_data.amount,
            percentage=rule_data.percentage,
            priority=rule_data.priority,
            name=rule_data.name,
            is_active=rule_data.is_active,
        )

        self.db.add(rule)
        await self.db.commit()
        await self.db.refresh(rule)

        logger.info(f"Created distribution rule {rule.id} for user {user_id}")
        return rule

    async def get_rule(self, rule_id: UUID, user_id: UUID) -> IncomeDistributionRule:
        """Get a distribution rule by ID."""
        result = await self.db.execute(
            select(IncomeDistributionRule).where(
                and_(
                    IncomeDistributionRule.id == rule_id,
                    IncomeDistributionRule.user_id == user_id,
                    IncomeDistributionRule.deleted_at.is_(None),
                )
            )
        )
        rule = result.scalar_one_or_none()
        if not rule:
            raise RuleNotFoundError(f"Rule {rule_id} not found")
        return rule

    async def get_rules(
        self,
        user_id: UUID,
        income_source_id: Optional[UUID] = None,
        is_active: Optional[bool] = None,
    ) -> List[IncomeDistributionRule]:
        """
        Get distribution rules for a user.

        Args:
            user_id: User UUID
            income_source_id: Optional filter by income source
            is_active: Optional filter by active status

        Returns:
            List of IncomeDistributionRule
        """
        query = select(IncomeDistributionRule).where(
            and_(
                IncomeDistributionRule.user_id == user_id,
                IncomeDistributionRule.deleted_at.is_(None),
            )
        )

        if income_source_id is not None:
            # Get rules for specific source OR global rules (null income_source_id)
            query = query.where(
                (IncomeDistributionRule.income_source_id == income_source_id) |
                (IncomeDistributionRule.income_source_id.is_(None))
            )

        if is_active is not None:
            query = query.where(IncomeDistributionRule.is_active == is_active)

        query = query.order_by(IncomeDistributionRule.priority)

        result = await self.db.execute(query)
        return list(result.scalars().all())

    async def update_rule(
        self,
        rule_id: UUID,
        user_id: UUID,
        rule_data: IncomeDistributionRuleUpdate,
    ) -> IncomeDistributionRule:
        """Update a distribution rule."""
        rule = await self.get_rule(rule_id, user_id)

        update_data = rule_data.model_dump(exclude_unset=True)

        # Validate targets if being updated
        if 'target_account_id' in update_data and update_data['target_account_id']:
            account = await self.db.get(SavingsAccount, update_data['target_account_id'])
            if not account or account.user_id != user_id:
                raise InvalidRuleError("Invalid target account")

        if 'target_goal_id' in update_data and update_data['target_goal_id']:
            goal = await self.db.get(Goal, update_data['target_goal_id'])
            if not goal or goal.user_id != user_id:
                raise InvalidRuleError("Invalid target goal")

        for key, value in update_data.items():
            setattr(rule, key, value)

        rule.updated_at = datetime.utcnow()
        await self.db.commit()
        await self.db.refresh(rule)

        logger.info(f"Updated distribution rule {rule_id}")
        return rule

    async def delete_rule(self, rule_id: UUID, user_id: UUID) -> bool:
        """Soft delete a distribution rule."""
        rule = await self.get_rule(rule_id, user_id)
        rule.deleted_at = datetime.utcnow()
        await self.db.commit()
        logger.info(f"Deleted distribution rule {rule_id}")
        return True

    async def preview_distribution(
        self,
        user_id: UUID,
        income_amount: Decimal,
        currency: str,
        income_source_id: Optional[UUID] = None,
    ) -> IncomeDistributionPreviewResponse:
        """
        Preview how income would be distributed without actually applying.

        Args:
            user_id: User UUID
            income_amount: Amount to distribute
            currency: Currency code
            income_source_id: Optional income source for source-specific rules

        Returns:
            IncomeDistributionPreviewResponse with distribution preview
        """
        rules = await self.get_rules(user_id, income_source_id, is_active=True)

        distributions: List[DistributionPreview] = []
        remaining = income_amount
        total_distributed = Decimal("0")

        for rule in rules:
            if remaining <= 0:
                break

            # Calculate distribution amount
            dist_amount = Decimal("0")
            if rule.distribution_type == DistributionType.PERCENTAGE:
                dist_amount = income_amount * (rule.percentage / Decimal("100"))
            elif rule.distribution_type == DistributionType.FIXED_AMOUNT:
                dist_amount = min(rule.amount, remaining)
            elif rule.distribution_type == DistributionType.REMAINDER:
                dist_amount = remaining

            if dist_amount <= 0:
                continue

            # Get target name
            target_type = "account" if rule.target_account_id else "goal"
            target_id = rule.target_account_id or rule.target_goal_id
            target_name = "Unknown"

            if rule.target_account_id:
                account = await self.db.get(SavingsAccount, rule.target_account_id)
                if account:
                    target_name = account.name
            elif rule.target_goal_id:
                goal = await self.db.get(Goal, rule.target_goal_id)
                if goal:
                    target_name = goal.name

            distributions.append(DistributionPreview(
                rule_id=rule.id,
                rule_name=rule.name,
                target_type=target_type,
                target_id=target_id,
                target_name=target_name,
                amount=dist_amount,
                currency=currency,
            ))

            total_distributed += dist_amount
            remaining -= dist_amount

        return IncomeDistributionPreviewResponse(
            income_amount=income_amount,
            currency=currency,
            distributions=distributions,
            remaining_amount=remaining,
            total_distributed=total_distributed,
        )

    async def apply_distribution(
        self,
        user_id: UUID,
        income_transaction_id: UUID,
    ) -> List[Tuple[UUID, Decimal]]:
        """
        Apply distribution rules to an income transaction.

        Args:
            user_id: User UUID
            income_transaction_id: Income transaction to distribute

        Returns:
            List of (account_transaction_id, amount) tuples for created deposits
        """
        # Get the income transaction
        result = await self.db.execute(
            select(IncomeTransaction).where(
                and_(
                    IncomeTransaction.id == income_transaction_id,
                    IncomeTransaction.user_id == user_id,
                )
            )
        )
        transaction = result.scalar_one_or_none()
        if not transaction:
            raise DistributionServiceError("Income transaction not found")

        # Preview distribution
        preview = await self.preview_distribution(
            user_id,
            transaction.amount,
            transaction.currency,
            transaction.source_id,
        )

        # Apply each distribution
        transaction_service = TransactionService(self.db)
        created_deposits: List[Tuple[UUID, Decimal]] = []

        for dist in preview.distributions:
            if dist.target_type == "account":
                try:
                    # Create deposit to account
                    account_txn = await transaction_service.create_deposit(
                        account_id=dist.target_id,
                        user_id=user_id,
                        amount=dist.amount,
                        source_type="income",
                        source_id=income_transaction_id,
                        description=f"Income distribution: {dist.rule_name or 'Auto-distribution'}",
                    )
                    created_deposits.append((account_txn.id, dist.amount))
                except Exception as e:
                    logger.error(f"Failed to apply distribution to account {dist.target_id}: {e}")

            elif dist.target_type == "goal":
                try:
                    # Get the goal and update its current_amount
                    goal_result = await self.db.execute(
                        select(Goal).where(
                            and_(
                                Goal.id == dist.target_id,
                                Goal.user_id == user_id,
                                Goal.is_active == True
                            )
                        )
                    )
                    goal = goal_result.scalar_one_or_none()

                    if goal:
                        # Add the distribution amount to the goal's current amount
                        goal.current_amount = Decimal(str(goal.current_amount)) + dist.amount

                        # Recalculate progress percentage
                        if goal.target_amount and goal.target_amount > 0:
                            goal.progress_percentage = (goal.current_amount / goal.target_amount) * 100

                        # Check for completion
                        if goal.current_amount >= goal.target_amount and not goal.is_completed:
                            goal.is_completed = True
                            goal.completed_at = datetime.utcnow()

                        goal.updated_at = datetime.utcnow()

                        # Record progress snapshot
                        from app.modules.goals.service import record_progress_snapshot
                        await record_progress_snapshot(
                            self.db, user_id, goal.id,
                            current_amount=goal.current_amount,
                            trigger_type="transaction",
                            notes=f"Income distribution: {dist.rule_name or 'Auto-distribution'}"
                        )

                        logger.info(f"Distributed {dist.amount} to goal {dist.target_id}")
                    else:
                        logger.warning(f"Goal {dist.target_id} not found or inactive for distribution")
                except Exception as e:
                    logger.error(f"Failed to distribute to goal {dist.target_id}: {e}")

        # Update income transaction status if any deposits were made
        if created_deposits:
            transaction.status = IncomeTransactionStatus.DEPOSITED
            transaction.deposited_to_account_id = preview.distributions[0].target_id if preview.distributions else None
            if created_deposits:
                transaction.account_transaction_id = created_deposits[0][0]
            await self.db.commit()

        logger.info(f"Applied distribution for income transaction {income_transaction_id}: {len(created_deposits)} deposits")
        return created_deposits

    async def enrich_rule_response(
        self,
        rule: IncomeDistributionRule,
    ) -> IncomeDistributionRuleResponse:
        """
        Enrich a rule with related entity names for response.
        """
        response_data = {
            "id": rule.id,
            "user_id": rule.user_id,
            "income_source_id": rule.income_source_id,
            "target_account_id": rule.target_account_id,
            "target_goal_id": rule.target_goal_id,
            "distribution_type": rule.distribution_type,
            "amount": rule.amount,
            "percentage": rule.percentage,
            "priority": rule.priority,
            "name": rule.name,
            "is_active": rule.is_active,
            "created_at": rule.created_at,
            "updated_at": rule.updated_at,
        }

        # Get related entity names
        if rule.income_source_id:
            source = await self.db.get(IncomeSource, rule.income_source_id)
            if source:
                response_data["income_source_name"] = source.name

        if rule.target_account_id:
            account = await self.db.get(SavingsAccount, rule.target_account_id)
            if account:
                response_data["target_account_name"] = account.name

        if rule.target_goal_id:
            goal = await self.db.get(Goal, rule.target_goal_id)
            if goal:
                response_data["target_goal_name"] = goal.name

        return IncomeDistributionRuleResponse(**response_data)

File: backend-nest/src/modules/income/services/distribution.service.ts
import { Inject, Injectable } from '@nestjs/common';
import { InjectDataSource } from '@nestjs/typeorm';
import { DataSource } from 'typeorm';
import { ListResponse, listed } from '../../../common/dto/page-query.dto';
import {
  DetailException,
  NotFoundException,
} from '../../../common/exceptions/app.exception';
import {
  decAdd,
  decCmp,
  decDiv,
  decMin,
  decMul,
  decSub,
} from '../../../common/money/money';
import {
  OwnedRepository,
  ownedRepositoryToken,
} from '../../../common/repository/owned.repository';
import { Goal } from '../../goals/entities/goal.entity';
import { GoalProgressService } from '../../goals/goal-progress.service';
import { AccountTransactionService } from '../../savings/account-transaction.service';
import { SavingsAccount } from '../../savings/entities/savings-account.entity';
import {
  CreateDistributionRuleDto,
  DistributionPreviewQueryDto,
  UpdateDistributionRuleDto,
} from '../dto/distribution.dto';
import { ListDistributionRulesQueryDto } from '../dto/income-query.dto';
import { IncomeDistributionRule } from '../entities/income-distribution-rule.entity';
import { IncomeSource } from '../entities/income-source.entity';
import { IncomeTransaction } from '../entities/income-transaction.entity';
import { DISTRIBUTION_TYPE_TO_NAME } from '../enums';
import {
  IncomeDistributionRuleResponse,
  toRuleResponse,
} from '../mappers/income-response.mapper';

export interface DistributionPreviewItem {
  rule_id: string;
  rule_name: string | null;
  target_type: string;
  target_id: string;
  target_name: string;
  amount: string;
  currency: string;
}

export interface DistributionPreviewResponse {
  income_amount: string;
  currency: string;
  distributions: DistributionPreviewItem[];
  remaining_amount: string;
  total_distributed: string;
}

@Injectable()
export class DistributionService {
  constructor(
    @Inject(ownedRepositoryToken(IncomeDistributionRule))
    private readonly rules: OwnedRepository<IncomeDistributionRule>,
    @Inject(ownedRepositoryToken(IncomeSource))
    private readonly sources: OwnedRepository<IncomeSource>,
    @Inject(ownedRepositoryToken(IncomeTransaction))
    private readonly transactions: OwnedRepository<IncomeTransaction>,
    private readonly deposits: AccountTransactionService,
    private readonly goalProgress: GoalProgressService,
    @InjectDataSource() private readonly dataSource: DataSource,
  ) {}

  // ---------------------------------------------------------------- rules CRUD

  async list(
    userId: string,
    query: ListDistributionRulesQueryDto,
  ): Promise<ListResponse<IncomeDistributionRuleResponse>> {
    const rules = await this.findRules(userId, {
      incomeSourceId: query.income_source_id,
      isActive: query.is_active,
    });
    const enriched = await Promise.all(
      rules.map((rule) => this.enrich(userId, rule)),
    );
    // FastAPI's envelope here is {items, total} with total = len(items) — no page/page_size.
    return listed(enriched);
  }

  async get(
    userId: string,
    ruleId: string,
  ): Promise<IncomeDistributionRuleResponse> {
    return this.enrich(userId, await this.findOwnedOrFail(userId, ruleId));
  }

  async create(
    userId: string,
    dto: CreateDistributionRuleDto,
  ): Promise<IncomeDistributionRuleResponse> {
    // Validation order matters — FastAPI returns the first failure and clients read the message.
    if (!dto.target_account_id && !dto.target_goal_id) {
      throw new DetailException(
        400,
        'Rule must have either a target account or target goal',
      );
    }
    if (dto.distribution_type === 'percentage' && !dto.percentage) {
      throw new DetailException(
        400,
        'Percentage type requires percentage value',
      );
    }
    if (dto.distribution_type === 'fixed_amount' && !dto.amount) {
      throw new DetailException(400, 'Fixed amount type requires amount value');
    }
    await this.assertTargetsOwned(userId, dto);

    const rule = this.rules.raw.create({
      userId,
      incomeSourceId: dto.income_source_id ?? null,
      targetAccountId: dto.target_account_id ?? null,
      targetGoalId: dto.target_goal_id ?? null,
      distributionType: DISTRIBUTION_TYPE_TO_NAME[dto.distribution_type],
      amount: dto.amount ?? null,
      percentage: dto.percentage ?? null,
      priority: dto.priority,
      name: dto.name ?? null,
      isActive: dto.is_active,
    });
    await this.rules.raw.save(rule);
    return this.enrich(userId, rule);
  }

  async update(
    userId: string,
    ruleId: string,
    dto: UpdateDistributionRuleDto,
  ): Promise<IncomeDistributionRuleResponse> {
    const rule = await this.findOwnedOrFail(userId, ruleId);
    await this.assertTargetsOwned(userId, dto);

    if ('income_source_id' in dto) {
      rule.incomeSourceId = dto.income_source_id ?? null;
    }
    if ('target_account_id' in dto) {
      rule.targetAccountId = dto.target_account_id ?? null;
    }
    if ('target_goal_id' in dto) rule.targetGoalId = dto.target_goal_id ?? null;
    if (dto.distribution_type !== undefined) {
      rule.distributionType = DISTRIBUTION_TYPE_TO_NAME[dto.distribution_type];
    }
    if ('amount' in dto) rule.amount = dto.amount ?? null;
    if ('percentage' in dto) rule.percentage = dto.percentage ?? null;
    if (dto.priority !== undefined) rule.priority = dto.priority;
    if ('name' in dto) rule.name = dto.name ?? null;
    if (dto.is_active !== undefined) rule.isActive = dto.is_active;

    await this.rules.raw.save(rule);
    return this.enrich(userId, rule);
  }

  async remove(userId: string, ruleId: string): Promise<void> {
    const rule = await this.findOwnedOrFail(userId, ruleId);
    await this.rules.raw.softRemove(rule);
  }

  // ------------------------------------------------------------------- preview

  async preview(
    userId: string,
    query: DistributionPreviewQueryDto,
  ): Promise<DistributionPreviewResponse> {
    const distributions = await this.computeDistributions(
      userId,
      query.income_amount,
      query.currency,
      query.income_source_id ?? null,
    );
    return {
      income_amount: query.income_amount,
      currency: query.currency,
      ...distributions,
    };
  }

  /**
   * The core allocation loop. Three behaviours here look like bugs and are deliberate copies:
   *  - a percentage is taken off the ORIGINAL amount, never off what remains, so rules summing
   *    above 100% drive `remaining` negative (the loop then stops on the NEXT iteration, after the
   *    over-allocating rule has already been applied in full);
   *  - a fixed amount is clamped to what remains, so it can be partially satisfied;
   *  - nothing is rounded. The unrounded value is what gets written, and the numeric(12,2) column
   *    does the rounding in Postgres.
   */
  private async computeDistributions(
    userId: string,
    incomeAmount: string,
    currency: string,
    incomeSourceId: string | null,
  ): Promise<{
    distributions: DistributionPreviewItem[];
    remaining_amount: string;
    total_distributed: string;
  }> {
    const rules = await this.findRules(userId, {
      incomeSourceId: incomeSourceId ?? undefined,
      isActive: true,
    });

    const distributions: DistributionPreviewItem[] = [];
    let remaining = incomeAmount;
    let totalDistributed = '0';

    for (const rule of rules) {
      if (decCmp(remaining, '0') <= 0) break;

      let amount = '0';
      if (rule.distributionType === 'PERCENTAGE') {
        amount = decMul(incomeAmount, decDiv(rule.percentage ?? '0', '100'));
      } else if (rule.distributionType === 'FIXED_AMOUNT') {
        amount = decMin(rule.amount ?? '0', remaining);
      } else {
        amount = remaining;
      }
      if (decCmp(amount, '0') <= 0) continue;

      const targetType = rule.targetAccountId ? 'account' : 'goal';
      const targetId = (rule.targetAccountId ?? rule.targetGoalId)!;
      distributions.push({
        rule_id: rule.id,
        rule_name: rule.name,
        target_type: targetType,
        target_id: targetId,
        target_name: (await this.targetName(userId, rule)) ?? 'Unknown',
        amount,
        // Echoed verbatim from the request; FastAPI does not upper-case it here.
        currency,
      });

      totalDistributed = decAdd(totalDistributed, amount);
      remaining = decSub(remaining, amount);
    }

    return {
      distributions,
      remaining_amount: remaining,
      total_distributed: totalDistributed,
    };
  }

  // ------------------------------------------------------------------ applying

  async apply(
    userId: string,
    transactionId: string,
  ): Promise<{
    message: string;
    deposits: Array<{ account_transaction_id: string; amount: number }>;
  }> {
    // One transaction for the whole distribution. FastAPI commits per deposit and again per goal
    // snapshot, so a failure halfway leaves money moved into some targets and not others.
    return this.dataSource.transaction(async (manager) => {
      const transaction = await this.transactions
        .withManager(manager)
        .findOne(userId, { id: transactionId });
      if (!transaction) {
        throw new DetailException(400, 'Income transaction not found');
      }

      const { distributions } = await this.computeDistributions(
        userId,
        transaction.amount,
        transaction.currency,
        transaction.sourceId,
      );

      const created: Array<{ account_transaction_id: string; amount: string }> =
        [];
      let firstAccountTargetId: string | null = null;

      for (const distribution of distributions) {
        const label = `Income distribution: ${distribution.rule_name ?? 'Auto-distribution'}`;
        if (distribution.target_type === 'account') {
          const accountTransaction = await this.deposits.createDeposit(
            manager,
            {
              accountId: distribution.target_id,
              userId,
              amount: distribution.amount,
              sourceType: 'income',
              sourceId: transaction.id,
              description: label,
            },
          );
          created.push({
            account_transaction_id: accountTransaction.id,
            amount: distribution.amount,
          });
          firstAccountTargetId ??= distribution.target_id;
        } else {
          await this.goalProgress.addToGoal(manager, {
            userId,
            goalId: distribution.target_id,
            amount: distribution.amount,
            notes: label,
          });
        }
      }

      if (created.length > 0) {
        transaction.status = 'DEPOSITED';
        // FastAPI writes distributions[0].target_id here, which is a GOAL id whenever a goal rule
        // sorts first — into a column whose FK points at savings_accounts, i.e. a guaranteed
        // violation. Use the account that was actually credited.
        transaction.depositedToAccountId = firstAccountTargetId;
        transaction.accountTransactionId = created[0].account_transaction_id;
        await manager.save(transaction);
      }

      return {
        message: `Successfully distributed income to ${created.length} account(s)`,
        // The one place in this module where an amount is a JSON number, not a string: FastAPI
        // builds this dict by hand with float(amount) and response_model=dict, so nothing coerces
        // it back to a Decimal.
        deposits: created.map((deposit) => ({
          account_transaction_id: deposit.account_transaction_id,
          amount: Number(deposit.amount),
        })),
      };
    });
  }

  // ------------------------------------------------------------------- helpers

  private async findOwnedOrFail(
    userId: string,
    ruleId: string,
  ): Promise<IncomeDistributionRule> {
    const rule = await this.rules.findOne(userId, { id: ruleId });
    if (!rule) throw new NotFoundException('Distribution rule not found');
    return rule;
  }

  /**
   * When an income source is given, global rules (income_source_id IS NULL) apply as well — so a
   * source-specific query is an OR, not an equality.
   */
  private findRules(
    userId: string,
    filters: { incomeSourceId?: string; isActive?: boolean },
  ): Promise<IncomeDistributionRule[]> {
    const builder = this.rules.qb(userId, 'r');
    if (filters.incomeSourceId) {
      builder.andWhere(
        '(r.income_source_id = :sourceId OR r.income_source_id IS NULL)',
        { sourceId: filters.incomeSourceId },
      );
    }
    if (filters.isActive !== undefined) {
      builder.andWhere('r.is_active = :isActive', {
        isActive: filters.isActive,
      });
    }
    // ORDER BY priority only — FastAPI has no secondary sort, so ties are DB order.
    return builder.orderBy('r.priority', 'ASC').getMany();
  }

  /**
   * Every lookup is user-scoped. FastAPI's enrich_rule_response fetches by primary key with no
   * owner check, and its update path never validates income_source_id ownership — so a user can
   * point a rule at someone else's source and read that source's name back out. Fixed here.
   */
  private async enrich(
    userId: string,
    rule: IncomeDistributionRule,
  ): Promise<IncomeDistributionRuleResponse> {
    const [incomeSourceName, targetAccountName, targetGoalName] =
      await Promise.all([
        rule.incomeSourceId
          ? this.sources
              .findOne(userId, { id: rule.incomeSourceId })
              .then((source) => source?.name ?? null)
          : Promise.resolve(null),
        rule.targetAccountId
          ? this.dataSource
              .getRepository(SavingsAccount)
              .findOne({ where: { id: rule.targetAccountId, userId } })
              .then((account) => account?.name ?? null)
          : Promise.resolve(null),
        rule.targetGoalId
          ? this.dataSource
              .getRepository(Goal)
              .findOne({ where: { id: rule.targetGoalId, userId } })
              .then((goal) => goal?.name ?? null)
          : Promise.resolve(null),
      ]);

    return toRuleResponse(rule, {
      incomeSourceName,
      targetAccountName,
      targetGoalName,
    });
  }

  private async targetName(
    userId: string,
    rule: IncomeDistributionRule,
  ): Promise<string | null> {
    if (rule.targetAccountId) {
      const account = await this.dataSource
        .getRepository(SavingsAccount)
        .findOne({ where: { id: rule.targetAccountId, userId } });
      return account?.name ?? null;
    }
    if (rule.targetGoalId) {
      const goal = await this.dataSource
        .getRepository(Goal)
        .findOne({ where: { id: rule.targetGoalId, userId } });
      return goal?.name ?? null;
    }
    return null;
  }

  /** Ownership checks for the three foreign keys a rule can carry. */
  private async assertTargetsOwned(
    userId: string,
    dto: CreateDistributionRuleDto | UpdateDistributionRuleDto,
  ): Promise<void> {
    if (dto.income_source_id) {
      const source = await this.sources.findOne(userId, {
        id: dto.income_source_id,
      });
      if (!source) throw new DetailException(400, 'Invalid income source');
    }
    if (dto.target_account_id) {
      const account = await this.dataSource
        .getRepository(SavingsAccount)
        .findOne({ where: { id: dto.target_account_id, userId } });
      if (!account) throw new DetailException(400, 'Invalid target account');
    }
    if (dto.target_goal_id) {
      const goal = await this.dataSource
        .getRepository(Goal)
        .findOne({ where: { id: dto.target_goal_id, userId } });
      if (!goal) throw new DetailException(400, 'Invalid target goal');
    }
  }
}
